### PR TITLE
Convert MomentumMassBDF2NodeSuppAlg to NGP

### PR DESCRIPTION
**Pull request by jrood-nrel**

### CODE REVIEW

File: src/node_kernels/MomentumMassBDFNodeKernel.C [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    const NodeKernelTraits::DblType uNp1       = velocityNp1_.get(node, i);
+    const NodeKernelTraits::DblType rhoNm1     = densityNm1_.get(node, i);
+    const NodeKernelTraits::DblType rhoN       = densityN_.get(node, i);
+    const NodeKernelTraits::DblType rhoNp1     = densityNp1_.get(node, i);
+    const NodeKernelTraits::DblType dualVolume = dualNodalVolume_.get(node, i);
+    const NodeKernelTraits::DblType dpdx       = dpdx_.get(node, i);
+    const NodeKernelTraits::DblType lhsfac = gamma1_*rhoNp1*dualVolume/dt_;
+
+    rhs(i) += -(gamma1_*rhoNp1*uNp1 + gamma2_*rhoN*uN + gamma3_*rhoNm1*uNm1)*dualVolume/dt_ - dpdx*dualVolume;
+    const int row = i*nDim;
```
Comment (sayerhs) [inline src/node_kernels/MomentumMassBDFNodeKernel.C]: `lhs` is a `3x3` matrix for `NodeKernel` this indexing is incorrect. Next line should be just `lhs(i, i)`.

File: unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+#include "node_kernels/MomentumMassBDFNodeKernel.h"
+
+#ifndef KOKKOS_ENABLE_CUDA
+namespace {
+namespace bdf_golds {
+namespace momentum_mass {
+
+static constexpr double lhs[24][24] = {
+{0.10943331816113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
```
Comment (sayerhs) [inline unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C]: This value should be `1.25`

File: unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
+{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
+{0, 0, 0, -0.12850080171032, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
```
Comment (sayerhs) [inline unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C]: Cannot have negative values in the diagonal from the time derivative term.

File: unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+static constexpr double rhs[24] =
+{-0.21886663632226, -0.25700160342063, -0.25700160342063, -0.25700160342063, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.21886663632226, -0.21886663632226, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, 0, -0.25700160342063, 0, 0, };
+
+} // momentum_mass
+} // bdf_golds
+} // anonymous namespace
+#endif
+
+TEST_F(MixtureFractionKernelHex8Mesh, NGP_momentum_mass_node)
```
Comment (sayerhs) [inline unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C]: This should be `MomentumKernelHex8Mesh`

File: unit_tests/kernels/UnitTestKernelUtils.h [hunk 5]
```diff
@@ -657,6 +657,7 @@ class MixtureFractionKernelHex8Mesh : public TestKernelHex8Mesh
     massFlowRate_(&meta_.declare_field<GenericFieldType>(stk::topology::ELEM_RANK,
                                                          "mass_flow_rate_scs")),
     dzdx_(&meta_.declare_field<VectorFieldType>(stk::topology::NODE_RANK, "dzdx")),
+    dpdx_(&meta_.declare_field<VectorFieldType>(stk::topology::NODE_RANK, "dpdx",2)),
```
Comment (sayerhs) [inline unit_tests/kernels/UnitTestKernelUtils.h]: Don't need this once the unit test is fixed to use `MomentumKernelHex8Mesh`

File: unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+  timeIntegrator.gamma1_ = 1.0;
+  timeIntegrator.gamma2_ = -1.0;
+  timeIntegrator.gamma3_ = 0.0;
+
+  unit_test_utils::NodeHelperObjects helperObjs(
+    bulk_, stk::topology::HEX_8, nDofs, partVec_[0]);
+
+  helperObjs.realm.timeIntegrator_ = &timeIntegrator;
+
+  helperObjs.nodeAlg->add_kernel<sierra::nalu::MomentumMassBDFNodeKernel>(bulk_, mixFraction_);
```
Comment (sayerhs) [inline unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C]: `Momentum` kernel constructor should not need to take an a scalar field argument. We assume by default that `velocity` exists and will be accessible internally.

File: src/node_kernels/MomentumMassBDFNodeKernel.C [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+MomentumMassBDFNodeKernel::MomentumMassBDFNodeKernel(
+  const stk::mesh::BulkData& bulk,
+  ScalarFieldType *scalarQ
+) : NGPNodeKernel<MomentumMassBDFNodeKernel>(),
+    nDim_(bulk.mesh_meta_data().spatial_dimension())
+{
+  const auto& meta = bulk.mesh_meta_data();
+
+  velocityNID_ = scalarQ->field_of_state(stk::mesh::StateN).mesh_meta_data_ordinal();
```
Comment (sayerhs) [inline src/node_kernels/MomentumMassBDFNodeKernel.C]: This is incorrect, `velocity` is a `VectorFieldType` and this line should be 

```
velocityNID_ = get_field_ordinal(meta, "velocity", stk::mesh::StateN);
```

Similarly for the remaining states.

**Comment by alanw0** [comment]
@sayerhs thanks for looking at these, and catching some problems. When doing these conversions, would it be possible for us to first unit-test the un-converted node-supp-alg to get the gold-values, then convert to the node-kernel and check against the gold-values. Clearly getting gold values from a newly-converted node-kernel doesn't verify anything.

But perhaps unit-testing the existing node-supp-algs requires some new helper-objects that use un-converted edge-algorithms?

**Comment by sayerhs** [comment]
@alanw0 We can use the unit tests in [algorithm](https://github.com/Exawind/nalu-wind/blob/master/unit_tests/algorithms/UnitTestMomentumBoussinesqSrcNodeSuppAlg.C) as a template for generating unit tests for the old algorithms and extract the gold values. I don't think this will require any new helper objects.

**Comment by sayerhs** [comment]
@alanw0 I think we can also use the lumped version of [MomentumMassElemKernel](https://github.com/Exawind/nalu-wind/blob/master/unit_tests/kernels/UnitTestMomentumMassElem.C#L193), as the node source supp alg should produce the same LHS/RHS as the lumped-mass version of the element kernel.

**Comment by sayerhs** [comment]
@alanw0 @jrood-nrel Here are the fixes for the `MomentumMassBDFNodeKernel`, there were several incorrect out of bounds array access in the `execute()` method that were causing the issues. It also required a change to use `MomentumKernelHex8Mesh` instead of `MixtureFractionHex8Mesh`. You'll notice that with these fixes, the `rhs` matches with the gold values for `NGP_momentum_time_derivative_lumped` (with the necessary node re-ordering). 

```diff
diff --git a/include/node_kernels/MomentumMassBDFNodeKernel.h b/include/node_kernels/MomentumMassBDFNodeKernel.h
index 7e89169..b23442d 100644
--- a/include/node_kernels/MomentumMassBDFNodeKernel.h
+++ b/include/node_kernels/MomentumMassBDFNodeKernel.h
@@ -24,8 +24,7 @@ class MomentumMassBDFNodeKernel : public NGPNodeKernel<MomentumMassBDFNodeKernel
 {
 public:
   MomentumMassBDFNodeKernel(
-    const stk::mesh::BulkData&,
-    ScalarFieldType*);
+    const stk::mesh::BulkData&);
 
   KOKKOS_FUNCTION
   MomentumMassBDFNodeKernel() = default;
diff --git a/src/node_kernels/MomentumMassBDFNodeKernel.C b/src/node_kernels/MomentumMassBDFNodeKernel.C
index 42aef28..7a56605 100644
--- a/src/node_kernels/MomentumMassBDFNodeKernel.C
+++ b/src/node_kernels/MomentumMassBDFNodeKernel.C
@@ -16,25 +16,27 @@ namespace sierra{
 namespace nalu{
 
 MomentumMassBDFNodeKernel::MomentumMassBDFNodeKernel(
-  const stk::mesh::BulkData& bulk,
-  ScalarFieldType *scalarQ
+  const stk::mesh::BulkData& bulk
 ) : NGPNodeKernel<MomentumMassBDFNodeKernel>(),
     nDim_(bulk.mesh_meta_data().spatial_dimension())
 {
   const auto& meta = bulk.mesh_meta_data();
 
-  velocityNID_ = scalarQ->field_of_state(stk::mesh::StateN).mesh_meta_data_ordinal();
+  const auto* velocity = meta.get_field<VectorFieldType>(
+      stk::topology::NODE_RANK, "velocity");
 
-  if (scalarQ->number_of_states() == 2)
+  velocityNID_ = velocity->field_of_state(stk::mesh::StateN).mesh_meta_data_ordinal();
+
+  if (velocity->number_of_states() == 2)
     velocityNm1ID_ = velocityNID_;
   else
-    velocityNm1ID_ = scalarQ->field_of_state(stk::mesh::StateNM1).mesh_meta_data_ordinal();
+    velocityNm1ID_ = velocity->field_of_state(stk::mesh::StateNM1).mesh_meta_data_ordinal();
 
-  velocityNp1ID_ = scalarQ->field_of_state(stk::mesh::StateNP1).mesh_meta_data_ordinal();
+  velocityNp1ID_ = velocity->field_of_state(stk::mesh::StateNP1).mesh_meta_data_ordinal();
 
   densityNID_ = get_field_ordinal(meta, "density", stk::mesh::StateN);
 
-  if (scalarQ->number_of_states() == 2)
+  if (velocity->number_of_states() == 2)
     densityNm1ID_ = densityNID_;
   else
     densityNm1ID_ = get_field_ordinal(meta, "density", stk::mesh::StateNM1);
@@ -71,21 +73,21 @@ MomentumMassBDFNodeKernel::execute(
 {
   const int nDim = nDim_;
 
+  const NodeKernelTraits::DblType rhoNm1     = densityNm1_.get(node, 0);
+  const NodeKernelTraits::DblType rhoN       = densityN_.get(node, 0);
+  const NodeKernelTraits::DblType rhoNp1     = densityNp1_.get(node, 0);
+  const NodeKernelTraits::DblType dualVolume = dualNodalVolume_.get(node, 0);
+
   // deal with lumped mass matrix (diagonal matrix)
   for ( int i = 0; i < nDim; ++i ) {
-    const NodeKernelTraits::DblType uNm1       = velocityNm1_.get(node, i);
-    const NodeKernelTraits::DblType uN         = velocityN_.get(node, i);
-    const NodeKernelTraits::DblType uNp1       = velocityNp1_.get(node, i);
-    const NodeKernelTraits::DblType rhoNm1     = densityNm1_.get(node, i);
-    const NodeKernelTraits::DblType rhoN       = densityN_.get(node, i);
-    const NodeKernelTraits::DblType rhoNp1     = densityNp1_.get(node, i);
-    const NodeKernelTraits::DblType dualVolume = dualNodalVolume_.get(node, i);
-    const NodeKernelTraits::DblType dpdx       = dpdx_.get(node, i);
+    const NodeKernelTraits::DblType uNm1   = velocityNm1_.get(node, i);
+    const NodeKernelTraits::DblType uN     = velocityN_.get(node, i);
+    const NodeKernelTraits::DblType uNp1   = velocityNp1_.get(node, i);
+    const NodeKernelTraits::DblType dpdx   = dpdx_.get(node, i);
     const NodeKernelTraits::DblType lhsfac = gamma1_*rhoNp1*dualVolume/dt_;
 
     rhs(i) += -(gamma1_*rhoNp1*uNp1 + gamma2_*rhoN*uN + gamma3_*rhoNm1*uNm1)*dualVolume/dt_ - dpdx*dualVolume;
-    const int row = i*nDim;
-    lhs(row+i, row+i) += lhsfac;
+    lhs(i, i) += lhsfac;
   }
 }
 
diff --git a/unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C b/unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C
index b194911..fa30414 100644
--- a/unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C
+++ b/unit_tests/node_kernels/UnitTestMomentumMassBDFNodeKernel.C
@@ -16,42 +16,22 @@ namespace {
 namespace bdf_golds {
 namespace momentum_mass {
 
-static constexpr double lhs[24][24] = {
-{0.10943331816113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, -0.12850080171032, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, -0.12850080171032, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0.10943331816113, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -0.12850080171032, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.10943331816113, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0.10943331816113, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -0.12850080171032, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-{0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, },
-};
-
 static constexpr double rhs[24] =
-{-0.21886663632226, -0.25700160342063, -0.25700160342063, -0.25700160342063, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.21886663632226, -0.21886663632226, -0.21886663632226, -0.25700160342063, -0.21886663632226, -0.25700160342063, 0, -0.25700160342063, 0, 0, };
+{0, 0, 0,
+ -0.056021853088904, -1.0112712429687, 0,
+ 1.0112712429687, -0.056021853088904, 0,
+ 0.53838846959557, -0.65043217577338, 0,
+ 0, 0, 0,
+ -0.056021853088904, -1.0112712429687, 0,
+ 1.0112712429687, -0.056021853088904, 0,
+ 0.53838846959557, -0.65043217577338, 0, };
 
 } // momentum_mass
 } // bdf_golds
 } // anonymous namespace
 #endif
 
-TEST_F(MixtureFractionKernelHex8Mesh, NGP_momentum_mass_node)
+TEST_F(MomentumKernelHex8Mesh, NGP_momentum_mass_node)
 {
   // Only execute for 1 processor runs
   if (bulk_.parallel_size() > 1) return;
@@ -72,7 +52,7 @@ TEST_F(MixtureFractionKernelHex8Mesh, NGP_momentum_mass_node)
 
   helperObjs.realm.timeIntegrator_ = &timeIntegrator;
 
-  helperObjs.nodeAlg->add_kernel<sierra::nalu::MomentumMassBDFNodeKernel>(bulk_, mixFraction_);
+  helperObjs.nodeAlg->add_kernel<sierra::nalu::MomentumMassBDFNodeKernel>(bulk_);
 
   helperObjs.execute();
 
@@ -80,11 +60,18 @@ TEST_F(MixtureFractionKernelHex8Mesh, NGP_momentum_mass_node)
   EXPECT_EQ(helperObjs.linsys->lhs_.extent(0), 24u);
   EXPECT_EQ(helperObjs.linsys->lhs_.extent(1), 24u);
   EXPECT_EQ(helperObjs.linsys->rhs_.extent(0), 24u);
+  EXPECT_EQ(helperObjs.linsys->numSumIntoCalls_, 8);
 
-  namespace gold_values = bdf_golds::momentum_mass;
-  unit_test_kernel_utils::expect_all_near(helperObjs.linsys->rhs_, gold_values::rhs, 1.0e-14);
-  unit_test_kernel_utils::expect_all_near<24>(helperObjs.linsys->lhs_, gold_values::lhs, 1.0e-14);
+  // Exact LHS expected
+  std::vector<double> lhsExact(576, 0.0);
+  for (int i=0; i<24; i++)
+    lhsExact[i*24+i] = 1.25;
 
+  namespace gold_values = bdf_golds::momentum_mass;
+  unit_test_kernel_utils::expect_all_near(
+    helperObjs.linsys->rhs_, gold_values::rhs, 1.0e-12);
+  unit_test_kernel_utils::expect_all_near_2d(
+    helperObjs.linsys->lhs_, lhsExact.data());
 #endif
 
 }
```

**Comment by alanw0** [comment]
@sayerhs thanks, that is very helpful

**Comment by sayerhs** [comment]
Minor nitpick in my patch, the following line can be taken out of the `nDim` for-loop and computed once. 

```C++
     const NodeKernelTraits::DblType lhsfac = gamma1_*rhoNp1*dualVolume/dt_;
```

**Pull request merged**